### PR TITLE
UNet Flax with FlaxModelMixin

### PR DESCRIPTION
**Pull request by pcuenca**
This is an alternative to #485 that incorporates #493.

~~We'll probably close #485, but I'm having a strange bug where the configuration is not correctly applied. The json file is correctly read in `from_pretrained`, but the signature of the `__init__` method does not contain all the parameters by the time it's inspected in `extract_init_dict`.~~

Thanks to @mishig25 the aforementioned bug was resolved.

### CODE REVIEW

File: src/diffusers/models/unet_2d_condition_flax.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    in_channels=4
+    out_channels=4
+    down_block_types=("CrossAttnDownBlock2D", "CrossAttnDownBlock2D", "CrossAttnDownBlock2D", "DownBlock2D")
+    up_block_types=("UpBlock2D", "CrossAttnUpBlock2D", "CrossAttnUpBlock2D", "CrossAttnUpBlock2D")
+    block_out_channels=(224, 448, 672, 896)
+    layers_per_block=2
+    attention_head_dim=8
+    cross_attention_dim=768
+    dropout=0.1
+    dtype: jnp.dtype = jnp.float32
```
Comment (mishig25) [inline src/diffusers/models/unet_2d_condition_flax.py:32]: >> I'm having a strange bug where the configuration is not correctly applied. The json file is correctly read in from_pretrained, but the signature of the __init__ method does not contain all the parameters by the time it's inspected in extract_init_dict

flax being flax, you need to provide typings

```suggestion
class FlaxUNet2DConditionModel(nn.Module, FlaxModelMixin, ConfigMixin):
    sample_size:int=32
    in_channels:int=4
    out_channels:int=4
    down_block_types:Tuple=("CrossAttnDownBlock2D", "CrossAttnDownBlock2D", "CrossAttnDownBlock2D", "DownBlock2D")
    up_block_types:Tuple=("UpBlock2D", "CrossAttnUpBlock2D", "CrossAttnUpBlock2D", "CrossAttnUpBlock2D")
    block_out_channels:Tuple=(224, 448, 672, 896)
    layers_per_block:int=2
    attention_head_dim:int=8
    cross_attention_dim:int=768
    dropout:float=0.1
    dtype: jnp.dtype = jnp.float32
```

Comment (pcuenca) [reply src/diffusers/models/unet_2d_condition_flax.py:32]: Of course 🤦 ! I'll try that, thanks a lot!

Comment (mishig25) [reply src/diffusers/models/unet_2d_condition_flax.py]: plese let me know if it fixe the issue, I haven't tested locally

Comment (pcuenca) [reply src/diffusers/models/unet_2d_condition_flax.py]: It worked like a charm, thanks again!

File: src/diffusers/models/attention_flax.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        head_size = self.heads
+        tensor = tensor.reshape(batch_size // head_size, head_size, seq_len, dim)
+        tensor = jnp.transpose(tensor, (0, 2, 1, 3))
+        tensor = tensor.reshape(batch_size // head_size, seq_len, dim * head_size)
+        return tensor
+
+    def __call__(self, hidden_states, context=None, deterministic=True):
+        context = hidden_states if context is None else context
+
+        q = self.to_q(hidden_states)
```
Comment (pcuenca) [inline src/diffusers/models/attention_flax.py:55]: I will rename these variables to match the changes in the PyTorch version

File: src/diffusers/models/unet_2d_condition_flax.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    FlaxUpBlock2D,
+)
+
+
+# from flax.core.frozen_dict import FrozenDict
+
+
+
+@flax_register_to_config
+class FlaxUNet2DConditionModel(nn.Module, FlaxModelMixin, ConfigMixin):
```
Comment (pcuenca) [inline src/diffusers/models/unet_2d_condition_flax.py:23]: I'll add docstrings here

File: src/diffusers/models/unet_2d_condition_flax.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        self.conv_norm_out = nn.GroupNorm(num_groups=32, epsilon=1e-5)
+        self.conv_out = nn.Conv(
+            self.out_channels,
+            kernel_size=(3, 3),
+            strides=(1, 1),
+            padding=((1, 1), (1, 1)),
+            dtype=self.dtype,
+        )
+
+    def __call__(
```
Comment (pcuenca) [inline src/diffusers/models/unet_2d_condition_flax.py:156]: Docstrings also needed here

File: src/diffusers/models/unet_2d_condition_flax.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    down_block_types: Tuple = ("CrossAttnDownBlock2D", "CrossAttnDownBlock2D", "CrossAttnDownBlock2D", "DownBlock2D")
+    up_block_types: Tuple = ("UpBlock2D", "CrossAttnUpBlock2D", "CrossAttnUpBlock2D", "CrossAttnUpBlock2D")
+    block_out_channels: Tuple = (224, 448, 672, 896)
+    layers_per_block: int = 2
+    attention_head_dim: int = 8
+    cross_attention_dim: int = 768
+    dropout: float = 0.1
+    dtype: jnp.dtype = jnp.float32
+
+    # model args
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_condition_flax.py:36]: Think we can delete this comments :-)

File: src/diffusers/models/unet_2d_condition_flax.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            padding=((1, 1), (1, 1)),
+            dtype=self.dtype,
+        )
+
+    def __call__(
+        self,
+        sample,
+        timesteps,
+        encoder_hidden_states,
+        # params: dict = None,
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_condition_flax.py:161]: let's delete those :-)

File: src/diffusers/models/unet_2d_condition_flax.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+        # dropout_rng: jax.random.PRNGKey = None,
+        train: bool = False,
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_condition_flax.py:163]: train should trickle down here to all modules and all `deterministic` inputs for dropout should be set to `not train` (as it's done in Transformers: https://github.com/huggingface/transformers/blob/16913b3c9215f592f1240511f8da271dc07c3552/src/transformers/models/bert/modeling_flax_bert.py#L660

**Review by patrickvonplaten — APPROVED** [review]
Looks very nice! Think it's just about deleting some dead code and maybe fix the dropouts everywhere :-)

**Comment by pcuenca** [comment]
Changed since last review:
- Trickle down `deterministic`
- Rename `q`, `k`, `v`
- Docstrings (for Flax UNet only)
- FlaxUNet2DConditionOutput
- `init_weights` (not yet in use)

File: src/diffusers/models/attention_flax.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        hidden_states = self.reshape_batch_dim_to_heads(hidden_states)
+        hidden_states = self.proj_attn(hidden_states)
+        return hidden_states
+
+
+class FlaxBasicTransformerBlock(nn.Module):
+    dim: int
+    n_heads: int
+    d_head: int
+    dropout: float = 0.0
```
Comment (patil-suraj) [inline src/diffusers/models/attention_flax.py:80]: Let's add the dropout layer

File: src/diffusers/models/attention_flax.py [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+    dtype: jnp.dtype = jnp.float32
+
+    def setup(self):
+        # self attention
+        self.self_attn = FlaxAttentionBlock(self.dim, self.n_heads, self.d_head, self.dropout, dtype=self.dtype)
+        # cross attention
+        self.cross_attn = FlaxAttentionBlock(self.dim, self.n_heads, self.d_head, self.dropout, dtype=self.dtype)
```
Comment (patil-suraj) [inline src/diffusers/models/attention_flax.py:87]: The names should match with pt version for autoconversion.

```suggestion
        self.attn1 = FlaxAttentionBlock(self.dim, self.n_heads, self.d_head, self.dropout, dtype=self.dtype)
        # cross attention
        self.attn2 = FlaxAttentionBlock(self.dim, self.n_heads, self.d_head, self.dropout, dtype=self.dtype)
```

File: src/diffusers/models/attention_flax.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+            inner_dim,
+            kernel_size=(1, 1),
+            strides=(1, 1),
+            padding="VALID",
+            dtype=self.dtype,
+        )
+
+    def __call__(self, hidden_states, context, deterministic=True):
+        batch, height, width, channels = hidden_states.shape
+        # import ipdb; ipdb.set_trace()
```
Comment (patil-suraj) [inline src/diffusers/models/attention_flax.py:147]:
```suggestion
```

File: src/diffusers/models/attention_flax.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+
+        hidden_states = hidden_states.reshape(batch, height, width, channels)
+
+        hidden_states = self.proj_out(hidden_states)
+        hidden_states = hidden_states + residual
+
+        return hidden_states
+
+
+class FlaxGluFeedForward(nn.Module):
```
Comment (patil-suraj) [inline src/diffusers/models/attention_flax.py:165]: We will have to split this in two modules `FeedForward` and `GEGLU` like in PyTorch.

File: src/diffusers/models/attention_flax.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+    dim: int
+    dropout: float = 0.0
```
Comment (patil-suraj) [inline src/diffusers/models/attention_flax.py:167]: Let's add the `dropout` layer.

File: src/diffusers/models/embeddings_flax.py [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import math
+
+import flax.linen as nn
+import jax.numpy as jnp
+
+
+# This is like models.embeddings.get_timestep_embedding (PyTorch) but
+# less general (only handles the case we currently need).
```
Comment (patil-suraj) [inline src/diffusers/models/embeddings_flax.py:21]: Yes, we could update this once we start converting other models.

File: src/diffusers/models/resnet_flax.py [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+            self.out_channels,
+            kernel_size=(3, 3),
+            strides=(2, 2),
+            padding=((1, 1), (1, 1)),  # padding="VALID",
+            dtype=self.dtype,
+        )
+
+    def __call__(self, hidden_states):
+        # pad = ((0, 0), (0, 1), (0, 1), (0, 0))  # pad height and width dim
+        # hidden_states = jnp.pad(hidden_states, pad_width=pad)
```
Comment (patil-suraj) [inline src/diffusers/models/resnet_flax.py:45]:
```suggestion
```

File: src/diffusers/models/resnet_flax.py [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+            )
+
+    def __call__(self, hidden_states, temb, deterministic=True):
+        residual = hidden_states
+        hidden_states = self.norm1(hidden_states)
+        hidden_states = nn.swish(hidden_states)
+        hidden_states = self.conv1(hidden_states)
+
+        temb = self.time_emb_proj(nn.swish(temb))
+        temb = jnp.expand_dims(jnp.expand_dims(temb, 1), 1)
```
Comment (patil-suraj) [inline src/diffusers/models/resnet_flax.py:100]: (nit) let's add a comment here about the shapes, bit hard to understand with the code

File: src/diffusers/models/unet_2d_condition_flax.py [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+                Use deterministic functions and disable dropout when not training.
+
+        Returns:
+            [`~models.unet_2d_condition_flax.FlaxUNet2DConditionOutput`] or `tuple`:
+            [`~models.unet_2d_condition_flax.FlaxUNet2DConditionOutput`] if `return_dict` is True, otherwise a `tuple`.
+            When returning a tuple, the first element is the sample tensor.
+        """
+        # 1. time
+        t_emb = self.time_proj(timesteps)
+        t_emb = self.time_embedding(t_emb)
```
Comment (patil-suraj) [inline src/diffusers/models/unet_2d_condition_flax.py:218]: This expects that `timestpes` is an `array`, might not work if it's a scaler. We should check this and hanlde `scaler` to `array` conversion.

File: src/diffusers/models/unet_blocks_flax.py [hunk 17]
```diff
@@ -0,0 +1,263 @@
...
+                depth=1,
+                dtype=self.dtype,
+            )
+            attentions.append(attn_block)
+
+        self.resnets = resnets
+        self.attentions = attentions
+
+        if self.add_downsample:
+            self.downsample = FlaxDownsample2D(self.out_channels, dtype=self.dtype)
```
Comment (patil-suraj) [inline src/diffusers/models/unet_blocks_flax.py:58]: This should be a `list`, same as in PT

File: src/diffusers/models/unet_blocks_flax.py [hunk 18]
```diff
@@ -0,0 +1,263 @@
...
+                in_channels=in_channels,
+                out_channels=self.out_channels,
+                dropout_prob=self.dropout,
+                dtype=self.dtype,
+            )
+            resnets.append(res_block)
+        self.resnets = resnets
+
+        if self.add_downsample:
+            self.downsample = FlaxDownsample2D(self.out_channels, dtype=self.dtype)
```
Comment (patil-suraj) [inline src/diffusers/models/unet_blocks_flax.py:99]: same comment as above.

File: src/diffusers/models/unet_blocks_flax.py [hunk 19]
```diff
@@ -0,0 +1,263 @@
...
+                depth=1,
+                dtype=self.dtype,
+            )
+            attentions.append(attn_block)
+
+        self.resnets = resnets
+        self.attentions = attentions
+
+        if self.add_upsample:
+            self.upsample = FlaxUpsample2D(self.out_channels, dtype=self.dtype)
```
Comment (patil-suraj) [inline src/diffusers/models/unet_blocks_flax.py:154]: same comment as above.

File: src/diffusers/models/unet_blocks_flax.py [hunk 20]
```diff
@@ -0,0 +1,263 @@
...
+                out_channels=self.out_channels,
+                dropout_prob=self.dropout,
+                dtype=self.dtype,
+            )
+            resnets.append(res_block)
+
+        self.resnets = resnets
+
+        if self.add_upsample:
+            self.upsample = FlaxUpsample2D(self.out_channels, dtype=self.dtype)
```
Comment (patil-suraj) [inline src/diffusers/models/unet_blocks_flax.py:199]: same comment as above.

**Review by patil-suraj — CHANGES_REQUESTED** [review]
Looks good in general! Left some comments, more specifically

- We should add the `dropout` layers, I didn't add it in the original repo as it was just for inference.
- Make sure that the module and weight names match 1:1 with PyTorch. This is required as we need to provide interoperability with PT and flax models.

**Comment by patrickvonplaten** [comment]
Thanks a lot for the review here @patil-suraj - you're 100% right here. To move fast, I'd say we merge this PR though and solve the conversion/weigth naming in a new PR (opening an issue for this) as well as the dropout layers.

As discussed offline, feel free to merge @pcuenca and we'll adapt in a future PR according to @patil-suraj's comments here :-)

File: src/diffusers/models/attention_flax.py [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+    def setup(self):
+        inner_dim = self.dim_head * self.heads
+        self.scale = self.dim_head**-0.5
+
+        # Weights were exported with old names {to_q, to_k, to_v, to_out}
+        self.query = nn.Dense(inner_dim, use_bias=False, dtype=self.dtype, name="to_q")
+        self.key = nn.Dense(inner_dim, use_bias=False, dtype=self.dtype, name="to_k")
+        self.value = nn.Dense(inner_dim, use_bias=False, dtype=self.dtype, name="to_v")
+
+        self.proj_attn = nn.Dense(self.query_dim, dtype=self.dtype, name="to_out")
```
[the remark this answers is not in the document]
Comment (pcuenca) [reply src/diffusers/models/attention_flax.py:35]: Yes, the original name was `self.to_q`, I changed it here to make it like the renamed PyTorch version but kept the same weight names.

**Comment by patrickvonplaten** [comment]
Opened to issues here for future PRs :-) 

- https://github.com/huggingface/diffusers/issues/523
- https://github.com/huggingface/diffusers/issues/522
